src/Data/Array/Polarized/Push.hs

```diff
@@ -11,55 +10,113 @@
+  )
+where
 
--- XXX: it might be better to hide the data constructor, in case we wish to
--- change the implementation.
-
-import Data.Array.Destination (DArray)
-import qualified Data.Array.Destination as DArray
 import qualified Data.Functor.Linear as Data
+import qualified Data.Array.Destination as DArray
+import Data.Array.Destination (DArray)
 import Data.Vector (Vector)
-import Prelude.Linear
 import qualified Prelude
+import qualified Unsafe.Linear as Unsafe
+import Prelude.Linear
+import GHC.Stack
+
 
--- TODO: the below isn't really true yet since no friendly way of constructing
--- a PushArray directly is given yet (see issue #62), but the spirit holds.
--- TODO: There's also a slight lie in that one might want to consume a
--- PushArray into a commutative monoid, for instance summing all the elements,
--- and this is not yet possible, although it should be.
+-- The Types
+-------------------------------------------------------------------------------
 
 -- | Push arrays are un-allocated finished arrays. These are finished
 -- computations passed along or enlarged until we are ready to allocate.
 data Array a where
+  Array :: (forall m. Monoid m => (a -> m) -> m) %1-> Array a
+  -- Developer notes:
+  --
+  -- Think of @(a -> m)@ as something that writes an @a@ and think of
+  -- @((a -> m) -> m)@ as something that takes a way to write a single element
+  -- and writes and concatenates all elements.
+  --
+  -- Also, note that in this formulation we don't know the length beforehand.
+
+data ArrayWriter a where
+  ArrayWriter :: (DArray a %1-> ()) %1-> !Int -> ArrayWriter a
   -- The second parameter is the length of the @DArray@
-  Array :: (forall b. (a %1-> b) -> DArray b %1-> ()) %1-> Int -> Array a
-  deriving Prelude.Semigroup via NonLinear (Array a)
 
-instance Data.Functor Array where
-  fmap f (Array k n) = Array (\g dest -> k (g . f) dest) n
 
-instance Semigroup (Array a) where
-  (<>) = append
+-- API
+-------------------------------------------------------------------------------
 
--- XXX: the use of Vector in the type of alloc is temporary (see also
--- "Data.Array.Destination")
 -- | Convert a push array into a vector by allocating. This would be a common
 -- end to a computation using push and pull arrays.
 alloc :: Array a %1-> Vector a
-alloc (Array k n) = DArray.alloc n (k id)
+alloc (Array k) = allocArrayWriter $ k singletonWriter where
+  singletonWriter :: a -> ArrayWriter a
+  singletonWriter a = ArrayWriter (DArray.fill a) 1
+
+  allocArrayWriter :: ArrayWriter a %1-> Vector a
+  allocArrayWriter (ArrayWriter writer len) = DArray.alloc len writer
 
 -- | @`make` x n@ creates a constant push array of length @n@ in which every
 -- element is @x@.
-make :: a -> Int -> Array a
-make x n = Array (\k -> DArray.replicate (k x)) n
+make :: HasCallStack => a -> Int -> Array a
+make x n
+  | n < 0 = error "Making negative length push array"
+  | otherwise = Array (\makeA -> mconcat $ Prelude.replicate n (makeA x))
```

Comment:
Base has an [`stimes`](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Semigroup.html#v:stimes) function for this. Maybe we should have one too?

Reply:
`Array $ \makeA -> stimes (makeA x)` might be more efficient if the `stimes` implementation for `ArrayWriter` could be more efficient.  However, I don't want to change `Monoid` in this PR.

Reply:
As you wish, but I think it's fine to lazily add stuff to the library (though I'm sensitive to the argument that changing a type class is not a cheap action, and may need to be done carefully).

It was not really about efficiency (though it may be), but about clarity. We can do this in a separate PR.